Refactor Econet integration

vangorra opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
`econet.add_vacation` and `econet.delete_vacation` services have been removed as downstream dependency (`pyeconet`) removed this functionality.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Updating `pyeconet` dependency to fix authentication errors.
- Add config flows.
- Using push for data updates instead of poll.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #38510
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14510

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

vangorra committed:
Refresh Econet integration

vangorra committed:
Removing services.yaml file.

vangorra committed:
Fixing spelling mistake.

vangorra commented:
Ready for review. codespell is throwing errors for the package name of `from pyeconet.equipments.water_heater import WaterHeater, WaterHeaterOperationMode`. Anybody know how to ignore this import line for code spell?

appleguru commented:
Changes look good to me. Eagerly awaiting getting this merged.

appleguru commented:
<img width="393" alt="Screen Shot 2020-09-19 at 10 16 29 PM" src="https://user-images.githubusercontent.com/595829/93692887-b37cf180-fac6-11ea-9dc9-090cee2e36bf.png">

<img width="1091" alt="Screen Shot 2020-09-19 at 10 18 03 PM" src="https://user-images.githubusercontent.com/595829/93692893-c263a400-fac6-11ea-8949-2a5e38789253.png">

@vangorra am testing this as a custom component so far, and things are mostly working. A couple of notes/questions:

```Logger: homeassistant
Source: runner.py:115 
First occurred: 10:06:29 PM (2 occurrences) 
Last logged: 10:06:29 PM

Error doing job: Unclosed client session
Error doing job: Unclosed connector```

```Logger: custom_components.econet.water_heater
Source: custom_components/econet/water_heater.py:105 
Integration: Rheem EcoNET Water Products (documentation) 
First occurred: 10:06:29 PM (1 occurrences) 
Last logged: 10:06:29 PM

Invalid operation mode mapping. WaterHeaterOperationMode.ENERGY_SAVING doesn't map. Please report this.```


This one seems likely to be downstream in pyeconet CC @w1ll1am23 
```Logger: pyeconet.equipments.water_heater
Source: /usr/local/lib/python3.8/site-packages/pyeconet/equipments/water_heater.py:49 
First occurred: 10:06:29 PM (21 occurrences) 
Last logged: 10:06:29 PM

Unknown mode: [Heat Pump ]
Unknown mode: [Electric ]
Unknown mode: [Vacation ]```

Also, how do I set the credentials for this? I got an error message to remove them from my config (which I did), and now see an "Econet" tile in my Integrations tab, which has my email address but no place to set a password? It seems to work though, so maybe it is using a cached token or something?

w1ll1am23 commented:
@vangorra I am working on a fix for some of @appleguru's issues. 

There will be a new version of pyeconet `0.1.5`

Supporting vacation as a mode for some unit and as a different attribute on others is going to be tricky.

w1ll1am23 commented on homeassistant/components/econet/water_heater.py:
```diff
@@ -46,87 +50,48 @@
-    "Electric": STATE_ELECTRIC,
-    "Heat Pump": STATE_HEAT_PUMP,
+ECONET_STATE_TO_HA: Dict[WaterHeaterOperationMode, str] = {
+    WaterHeaterOperationMode.ENERGY_SAVER: STATE_ECO,
```
This should be ENERGY_SAVING

w1ll1am23 commented:
@vangorra if you are interested in helping maintain the pyeconet library let me know, I have been super busy at work the last 6 months which is why I haven't gotten around to opening this PR myself.

vangorra commented:
@w1ll1am23 Sure. I'm interested in working on the pyeconet library. 
In regards to vacation modes. While useful independent of Home Assistant, vacation modes are easily replaced by using HASS' automations and provide the user with greater control. I'm all for sticking with the basics and ignoring "features" that HASS handles on it's own.

w1ll1am23 commented:
@vangorra I agree that the vacation functionality can be recreated in Home Assistant. The big problem here is that apparently some units support vacation as a mode. How will Home Assistant handle this? What if the user has vacation mode selected in the app what will HA show? Kinda the same question, but if we enable vacation mode support by using what Home Assistant has built in for away mode we still have the issue of what to show in the operation mode selector.

appleguru commented:
My unit is one of the ones that has the new "vacation mode"; it shows up as a second "off" entry in the UI and throws an error like this in the log:
<img width="399" alt="Screen Shot 2020-09-22 at 10 28 35 PM" src="https://user-images.githubusercontent.com/595829/93957801-05ba4e80-fd23-11ea-8126-a4acde499fa1.png">

`
Logger: custom_components.econet.water_heater
Source: custom_components/econet/water_heater.py:105 
Integration: Rheem EcoNET Water Products (documentation) 
First occurred: 10:15:11 PM (1 occurrences) 
Last logged: 10:15:11 PM

Invalid operation mode mapping. WaterHeaterOperationMode.VACATION doesn't map. Please report this.
`

One other issue we're seeing is that we see a "This entity does not have a unique id" error, and no entities show up on the integrations page:

<img width="294" alt="Screen Shot 2020-09-22 at 10 27 18 PM" src="https://user-images.githubusercontent.com/595829/93957707-d3a8ec80-fd22-11ea-8888-5d0b0b564500.png">

<img width="1282" alt="Screen Shot 2020-09-22 at 6 04 02 PM" src="https://user-images.githubusercontent.com/595829/93957720-da376400-fd22-11ea-9646-54081442a16c.png">

adamurban commented:
@w1ll1am23 can you rename `equipments` to `equipment` downstream to fix the codespell check that is failing?

w1ll1am23 commented:
> @w1ll1am23 can you rename `equipments` to `equipment` downstream to fix the codespell check that is failing?

Yeah, I guess I can do that. It's named that because that's how their API names it. I'll try to do that later tonight.

w1ll1am23 commented:
Bump the version for pyeconet to 0.1.5 and rename equipments to equipment and you should be good to go.

appleguru commented:
> Bump the version for pyeconet to 0.1.5 and rename equipments to equipment and you should be good to go.

@vangorra can you make this change in your PR? Should fix the code spell issue.

vangorra commented:
Fixing the unique_id issue is trivial. Downstream API 0.1.5 is broken. api.py still references pyeconet.equipments.

Support for vacation mode is easy enough to add however pyeconet does not seem to support it yet.

w1ll1am23 commented:
@vangorra sorry about that. I fixed it in 0.1.6.

Vacation is supported as a mode on @appleguru's unit, once we get this MR added we can follow up with another to get that fixed.

vangorra committed:
Using proper entity unique id.
Using latest release of downstream dep.

vangorra commented:
Updated PR with latest API and fixed unique id issues.
The errors people see are caused by the downstream API, nothing we can in an Hass about that.

@appleguru Since your unit supports vacation mode, is it possible to submit the equipment data in this ticket (https://github.com/w1ll1am23/pyeconet/issues/11)? I have a less complex unit that supports vacation as well. Hopefully we'll get enough of a sample to write a good generic solution.

appleguru commented:
@vangorra added my data. Your latest changes look good to me/seem to be working. I think this is good to merge!

w1ll1am23 commented:
@vangorra are you planning to merge this? I think it is good to go other than the vacation issue which can be addressed later.

vangorra commented:
Almost ready. I'll have time soon

w1ll1am23 commented on homeassistant/components/econet/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+"""Config flow for Gogogate2."""
```
This comment is wrong

w1ll1am23 commented on homeassistant/components/econet/water_heater.py:
```diff
@@ -158,22 +138,18 @@ def temperature_unit(self):
     def device_state_attributes(self):
         """Return the optional device state attributes."""
```
Users that have been testing out my version of these changes might be expecting 

ATTR_WIFI_SIGNAL = "wifi_signal_strength"
ATTR_ALERT_COUNT = "alert_count"

To be present in the state attributes, you may want to add those in.

If you are feeling crazy, it would be a good idea to create a base EcoNet entity which has all of the common stuff in it so the next MR for HVAC is easier. But not a requirement of course.

adamurban commented:
Is there a reason this hasn’t been merged yet? Minor nits aside, the existing integration is 100% broken / does not work today, which this PR fixes.
Please merge ASAP; we can follow on with an additional bug fix/improvement PR as needed.

zombielinux commented:
I agree with @adamurban I've been manually making these merges as homeassistant gets updated and haven't had any issues whatsoever. 

Lets merge this now since it passes all the checks and has no conflicts.

Lopton commented:
Also agree on this. Existing integration is completely non-functional.

adamurban commented:
@vangorra Can you make the small fixes above and rename this PR to indicate it fixes a broken integration?

vangorra committed:
Merge branch 'dev' into econet_refresh

vangorra committed:
Updated based on PR feedback.

adamurban reviewed (approved):
LGTM

Lopton commented:
This is still needed

zombielinux commented:
It looks like all the tests have passed and we're just waiting on a dev with write access to approve the merge.

appleguru commented:
> It looks like all the tests have passed and we're just waiting on a dev with write access to approve the merge.

I don’t believe this is needed anymore now that https://github.com/home-assistant/core/pull/44427 has been merged.

w1ll1am23 commented:
Correct this is not needed. Everyone please stop commenting so it will get auto closed.
